src/pipeline/filter.c

```diff
@@ -0,0 +1,378 @@
+				*comp_part             = (struct filter_part_compiled_t){0};
+
+				comp_part->data_offset = 0;
+				comp_part->iov         = &filter->parts[*part_idx]->constant[j];
+				comp_part->data_len    = comp_part->iov->iov_len;
+				comp_part->filter_func = getd_func_ptrs[func_idx];
+			}
+		} else if (!strncmp(part_type, "DAOS_FILTER_DKEY", part_type_s)) {
+			comp_part->data_offset = filter->parts[*part_idx]->data_offset;
+			comp_part->data_len    = filter->parts[*part_idx]->data_len;
+			comp_part->filter_func = getd_func_ptrs[func_idx];
+		}
+		D_GOTO(exit, rc = 0);
+	}
+
+	nops                    = calc_num_operands(filter->parts, *part_idx);
+	comp_part->num_operands = nops;
+
+	/** recursive calls for function parameters */
+	idx                     = *part_idx;
+	for (i = 0; i < filter->parts[idx]->num_operands; i++) {
+		*comp_part_idx += 1;
+		*part_idx += 1;
+		rc = compile_filter(filter, comp_filter, part_idx, comp_part_idx, type, type_len);
+		if (rc != 0)
+			D_GOTO(exit, rc);
+	}
+
+	func_idx = calc_filterfunc_idx(filter->parts, idx);
+	type_idx = calc_type_nosize_idx(calc_type_idx(*type, *type_len));
+
+	if (func_idx < 47)
+		func_idx += type_idx;
+
+	comp_part->filter_func     = filter_func_ptrs[func_idx];
+	comp_part->idx_end_subtree = *comp_part_idx;
+exit:
+	return rc;
+}
+
+static int
+compile_filters(daos_filter_t **ftrs, uint32_t nftrs, struct filter_compiled_t *c_ftrs)
+{
+	uint32_t            part_idx;
+	uint32_t            comp_part_idx;
+	uint32_t            comp_num_parts;
+	uint32_t            i, j;
+	char	       *type;
+	size_t              type_len;
+	int                 rc = 0;
+	daos_filter_part_t *part;
+
+	for (i = 0; i < nftrs; i++) {
+		comp_num_parts = ftrs[i]->num_parts;
+
+		for (j = 0; j < ftrs[i]->num_parts; j++) {
+			part = ftrs[i]->parts[j];
+			if (!strncmp((char *)part->part_type.iov_buf, "DAOS_FILTER_CONST",
+				     part->part_type.iov_len))
+				comp_num_parts += part->num_constants - 1;
+		}
+
+		D_ALLOC_ARRAY(c_ftrs[i].parts, comp_num_parts);
+		if (c_ftrs[i].parts == NULL)
+			D_GOTO(exit, rc = -DER_NOMEM);
+
+		c_ftrs[i].num_parts = comp_num_parts;
+		part_idx            = 0;
+		comp_part_idx       = 0;
+		type                = NULL;
+		type_len            = 0;
+		rc = compile_filter(ftrs[i], &c_ftrs[i], &part_idx, &comp_part_idx, &type,
+				    &type_len);
+		if (rc != 0)
+			D_GOTO(exit, rc);
+	}
+exit:
+	return rc;
+}
+
+int
+pipeline_compile(daos_pipeline_t *pipe, struct pipeline_compiled_t *comp_pipe)
+{
+	int rc                      = 0;
+
+	comp_pipe->num_filters      = 0;
+	comp_pipe->filters          = NULL;
+	comp_pipe->num_aggr_filters = 0;
+	comp_pipe->aggr_filters     = NULL;
+
+	if (pipe->num_filters > 0) {
+		D_ALLOC_ARRAY(comp_pipe->filters, pipe->num_filters);
+		if (comp_pipe->filters == NULL)
+			D_GOTO(exit, rc = -DER_NOMEM);
+
+		comp_pipe->num_filters = pipe->num_filters;
+		rc = compile_filters(pipe->filters, pipe->num_filters, comp_pipe->filters);
+		if (rc != 0)
```

Comment:
on failure, comp_pipe->filters is leaked

Reply:
fixed